File: typescript/src/02-mcp-client.ts
/**
 * 02-mcp-client.ts
 *
 * Connect to Hermes Plant's Streamable HTTP MCP endpoint, list available
 * tools, and call one. MCP handles the protocol envelope; the underlying
 * tool calls still go through x402 for paid endpoints, so your wallet
 * authorizes each paid call.
 *
 *   WALLET_PRIVATE_KEY=0x... npm run example:mcp
 */
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StreamableHTTPClientTransport } from "@modelcontextprotocol/sdk/client/streamableHttp.js";

const MCP_URL = process.env.HERMES_MCP_URL ?? "https://hermesplant.com/mcp";

async function main() {
  const transport = new StreamableHTTPClientTransport(new URL(MCP_URL));
  const client = new Client(
    { name: "hermesplant-examples-mcp", version: "0.1.0" },
    { capabilities: {} },
  );

  await client.connect(transport);
  console.log("Connected:", client.getServerVersion());

  const tools = await client.listTools();
  console.log(`Tools (${tools.tools.length}):`);
  for (const tool of tools.tools) {
    console.log(`  - ${tool.name}: ${tool.description ?? ""}`);
  }

  // Try a free tool first, such as discovery, health, catalog, or llms.
  // Escalate to a paid tool only after wallet auth is wired into the runtime.
  const freeTool = tools.tools.find((t) =>
    /discover|health|catalog|llms/i.test(t.name),
  );
  if (freeTool) {
    console.log(`\nCalling ${freeTool.name} (free) ...`);
    const result = await client.callTool({ name: freeTool.name, arguments: {} });
    console.log("Result:", JSON.stringify(result, null, 2).slice(0, 800));
  }

  await client.close();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: python/02-mcp-client.py
"""
02-mcp-client.py

Connect to Hermes Plant's Streamable HTTP MCP endpoint, list tools, and call
a free one. Paid tools still require an x402 wallet; see 01-call-endpoint.py
for the payment flow.
"""
import asyncio
import json
import os

from mcp import ClientSession
from mcp.client.streamable_http import streamablehttp_client

MCP_URL = os.environ.get("HERMES_MCP_URL", "https://hermesplant.com/mcp")


async def main() -> None:
    async with streamablehttp_client(MCP_URL) as (read_stream, write_stream, _):
        async with ClientSession(read_stream, write_stream) as session:
            await session.initialize()
            print("Connected.")

            tools = await session.list_tools()
            print(f"\nTools ({len(tools.tools)}):")
            for tool in tools.tools:
                description = tool.description or ""
                print(f"  - {tool.name}: {description}")

            free_tool = next(
                (
                    t for t in tools.tools
                    if any(token in t.name.lower() for token in ("discover", "health", "catalog", "llms"))
                ),
                None,
            )
            if free_tool:
                print(f"\nCalling {free_tool.name} (free) ...")
                result = await session.call_tool(free_tool.name, {})
                print(json.dumps(result.model_dump(), indent=2)[:800])


if __name__ == "__main__":
    asyncio.run(main())
